enhancements/sig-architecture/70-spread-policy/README.md

```diff
@@ -0,0 +1,226 @@
+
+### Use Case 6: Spread with Ratio (Out of Scope)
+
+Sometimes, we may need to spread the workload with a specified ratio.
+Let us consider a scenario where most of the access to the workload is from the east coast of the US, even spread between us-east-1 and us-west-1 would be inappropriate.
+Instead, we want to place more workload on us-east-1 than us-west-1.
+Besides, we want to control such preferences by user-specified configurations.
+
+**Example.** We want six clusters to deploy our backend service.
+Moreover, we want 2/3 of them to be located in us-east-1 and 1/3 to be located in us-west-1 since there are more clients from the east coast of the US.
+
+## Goal
+Generally, we have two goals for the spread policy:
+1. Cover the full functionality in the Motivation section;
+2. Implement a high-performance and scalable mechanism that works for thousands of clusters efficiently.
+
+## Proposal
+
+### API Spec
+```go
+type PlacementSpec struct {
+	// ...
+
+	// SpreadPolicy defines how placement decision should be distributed among a
+	// set of clusters.
+	SpreadPolicy []SpreadPolicyTerm `json:"spreadPolicy,omitempty"`
+}
+
+// SpreadPolicyTerm defines a terminology to spread placement decisions
+type SpreadPolicyTerm struct {
+	// TopologyKey is either a label key or a cluster claim name of ManagedClusters
+	// +required
+	TopologyKey string `json:"topologyKey"`
+
+	// TopologyKeyType indicates the type of TopologyKey. It could be Label or Claim.
+	// +required
+	TopologyKeyType string `json:"topologyKeyType"`
+
+	// MaxSkew describes the degree to which the workload may be unevenly distributed.
+	// If omitted, the scheduler treat this term a soft constraint. Otherwise, the
+	// scheduler would not schedule if MaxSkew cannot be satisfied (i.e., a hard constraint).
+	MaxSkew *int32 `json:"maxSkew,omitempty"`
+
+	// Weight indicates the weight of the SpreadPolicyTerm. It is required only when multiple
+	// terms are used. To spread the workload, the scheduler takes the even constraints with
+	// higher weight into prior consideration.
+	Weight int32 `json:"weight,omitempty"`
+}
+```
+
+### API Example
+
+```yaml
+apiVersion: cluster.open-cluster-management.io/v1alpha1
+kind: Placement
+metadata:
+  name: demo-placement
+spec:
+  # numberOfClusters is required if spreadConstraints contains
+  # a constraint of Even type.
+  numberOfClusters: 4
+  # Prediates work before the spread constraints to filter
+  # out some undesired topologies.
+  predicates:
+  - # ...
+  # SpreadPolicy contain two terms, which work
+  # independently and jointly.
+  spreadPolicy:
+    # When multiple terms are used, the user should specify their weights.
+    # The scheduler will consider to first do the spread according to the
+    # terms with higher weight.
+    # In this example, the scheduler will first spread the workload
+    # by providers.
```

Comment:
could you explain what happens when there is a higher weight of soft constraint and lower weight of hard constraint? It seems that hard constraint will always have the highest weight?

Reply:
A spread policy term with `maxSkew` can be considered as both a soft and a hard constraint. The soft part does the *best-effort* spread.  The hard part prevents some extreme cases that violate `maxSkew`. The hard part *does* have the highest weight semantically. But here the `weight` field is given weight to the soft parts and terms without `maxSkew`.

Reply:
I am not quite sure about the meaning of soft and hard part. When a policy is set, it must be either hard or soft constraint, right? a policy cannot be both hard and soft. If maxSkew is not set, doest it mean it is a hard constraint, and setting weight for this policy is meaningless?

Reply:
In this proposal, when `maxSkew` is not set, it means `maxSkew` is infinity. If users want to set a hard limit to an spread policy term, they need to explicitly set  `maxSkew` like `maxSkew: 1`. This design is welcome for comments.

Now let's consider a spread policy term with a hard limit like `maxSkew: 10`. If we have two regions `ra`, `rb`, and each of them provides enough cl;usters. We want to select 11 clusters in total for our workload. The hard limit prevents a placement like `ra: 11 clusters, rb: 0 cluster`, but it doesn't try the best to give an even spread like `ra: 6 clusters, rb: 5 clusters`. The placement `ra: 10 clusters, rb: 1 cluster` is also valid for the hard limit. To give an even spread, we still need to *score* the clusters. Therefore, there still needs a weight for the even spread scores when summing them with other scores from prioritizers .